File: java-original/src/main/java/com/demo/productcatalog/service/S3StorageService.java
package com.demo.productcatalog.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.*;

import jakarta.annotation.PostConstruct;
import java.io.InputStream;
import java.net.URI;

@Service
public class S3StorageService {

    @Value("${aws.s3.bucket-name}")
    private String bucketName;

    @Value("${aws.s3.region}")
    private String region;

    @Value("${aws.access-key-id}")
    private String accessKeyId;

    @Value("${aws.secret-access-key}")
    private String secretAccessKey;

    @Value("${aws.s3.endpoint:#{null}}")
    private String endpoint;

    private S3Client s3Client;

    @PostConstruct
    public void init() {
        var credentialsProvider = StaticCredentialsProvider.create(
                AwsBasicCredentials.create(accessKeyId, secretAccessKey)
        );

        var builder = S3Client.builder()
                .region(Region.of(region))
                .credentialsProvider(credentialsProvider);

        // Support LocalStack for local development
        if (endpoint != null && !endpoint.isEmpty()) {
            builder.endpointOverride(URI.create(endpoint))
                   .forcePathStyle(true);
        }

        this.s3Client = builder.build();
    }

    public String uploadFile(String key, InputStream inputStream, long contentLength, String contentType) {
        PutObjectRequest putRequest = PutObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .contentType(contentType)
                .build();

        s3Client.putObject(putRequest, RequestBody.fromInputStream(inputStream, contentLength));

        return getFileUrl(key);
    }

    public byte[] downloadFile(String key) {
        GetObjectRequest getRequest = GetObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .build();

        return s3Client.getObjectAsBytes(getRequest).asByteArray();
    }

    public void deleteFile(String key) {
        DeleteObjectRequest deleteRequest = DeleteObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .build();

        s3Client.deleteObject(deleteRequest);
    }

    public String getFileUrl(String key) {
        if (endpoint != null && !endpoint.isEmpty()) {
            return String.format("%s/%s/%s", endpoint, bucketName, key);
        }
        return String.format("https://%s.s3.%s.amazonaws.com/%s", bucketName, region, key);
    }
}

File: dotnet-migrated/ProductCatalog.Api/Services/S3StorageService.cs
using Amazon.S3;
using Amazon.S3.Model;

namespace ProductCatalog.Api.Services;

public interface IS3StorageService
{
    Task<string> UploadFileAsync(string key, Stream inputStream, string contentType);
    Task<byte[]> DownloadFileAsync(string key);
    Task DeleteFileAsync(string key);
    string GetFileUrl(string key);
}

public class S3StorageService : IS3StorageService
{
    private readonly IAmazonS3 _s3Client;
    private readonly string _bucketName;
    private readonly string _region;
    private readonly string? _endpoint;

    public S3StorageService(IConfiguration configuration)
    {
        _bucketName = configuration["AWS:S3:BucketName"] ?? "product-catalog-images";
        _region = configuration["AWS:Region"] ?? "us-east-1";
        _endpoint = configuration["AWS:S3:Endpoint"];

        var accessKeyId = configuration["AWS:AccessKeyId"] ?? "test";
        var secretAccessKey = configuration["AWS:SecretAccessKey"] ?? "test";

        var config = new AmazonS3Config
        {
            RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(_region)
        };

        // Support LocalStack for local development
        if (!string.IsNullOrEmpty(_endpoint))
        {
            config.ServiceURL = _endpoint;
            config.ForcePathStyle = true;
        }

        _s3Client = new AmazonS3Client(accessKeyId, secretAccessKey, config);
    }

    public async Task<string> UploadFileAsync(string key, Stream inputStream, string contentType)
    {
        var putRequest = new PutObjectRequest
        {
            BucketName = _bucketName,
            Key = key,
            InputStream = inputStream,
            ContentType = contentType
        };

        await _s3Client.PutObjectAsync(putRequest);

        return GetFileUrl(key);
    }

    public async Task<byte[]> DownloadFileAsync(string key)
    {
        var getRequest = new GetObjectRequest
        {
            BucketName = _bucketName,
            Key = key
        };

        using var response = await _s3Client.GetObjectAsync(getRequest);
        using var memoryStream = new MemoryStream();
        await response.ResponseStream.CopyToAsync(memoryStream);
        return memoryStream.ToArray();
    }

    public async Task DeleteFileAsync(string key)
    {
        var deleteRequest = new DeleteObjectRequest
        {
            BucketName = _bucketName,
            Key = key
        };

        await _s3Client.DeleteObjectAsync(deleteRequest);
    }

    public string GetFileUrl(string key)
    {
        if (!string.IsNullOrEmpty(_endpoint))
        {
            return $"{_endpoint}/{_bucketName}/{key}";
        }
        return $"https://{_bucketName}.s3.{_region}.amazonaws.com/{key}";
    }
}
